subprojects/execution/src/main/java/org/gradle/internal/execution/impl/OutputsCleaner.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.execution.impl;
+
+import com.google.common.collect.Ordering;
+import org.gradle.internal.file.FileType;
+import org.gradle.internal.fingerprint.FileCollectionFingerprint;
+import org.gradle.internal.fingerprint.FileSystemLocationFingerprint;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.util.Map;
+import java.util.PriorityQueue;
+import java.util.function.Predicate;
+
+public class OutputsCleaner {
+    private final PriorityQueue<File> directoriesToDelete;
```

Comment:
Why use a `PriorityQueue` instead of an `ArrayDeque`?

Reply:
Since we need to delete the leaf directories first since those can cause other directories to become empty.

Reply:
Ah, so longer paths pop towards the top of the queue, and thus will be deleted before shorter ones that are potentially parents... clever!